### PR TITLE
feat: Implement  UV lock translator

### PR DESCRIPTION
**Pull request by ewianda**
Close  https://github.com/jvolkman/rules_pycross/issues/103

### CODE REVIEW

**Comment by ewianda** [comment]
@jvolkman Seems the format of uv.lock is pretty close to pdm.lock and can share most of the code. Not sure if you have thoughts already on how this could be implemented

**Comment by jvolkman** [comment]
Thanks for the PR. Looks like it was a decent amount of work!

I think I would prefer the `uv` translator to just be a copy+paste of the `pdm` code, modified as needed, rather than trying to abstract a common implementation. Unless there is some guarantee that `uv` is tracking `pdm` model changes going forward, copying and pasting the implementation better aligns with the fact that they are separate formats and may diverge. It would also make reviewing easier, as most changes will just be additive.

**Comment by ewianda** [comment]
Fair enough. Does this apply only to the uv_translator.py, or should I copy-paste the `.bzl` functions /files as well

**Comment by jvolkman** [comment]
> Fair enough. Does this apply only to the uv_translator.py, or should I copy-paste the `.bzl` functions /files as well

I would just do it all.

File: pycross/private/bzlmod/lock_import.bzl [hunk 1]
```diff
@@ -110,6 +111,8 @@ def _lock_import_impl(module_ctx):
             lock_model_structs[tag.repo] = lock_repo_model_pdm(**{attr: getattr(tag, attr) for attr in PDM_IMPORT_ATTRS})
         for tag in module.tags.import_poetry:
             lock_model_structs[tag.repo] = lock_repo_model_poetry(**{attr: getattr(tag, attr) for attr in POETRY_IMPORT_ATTRS})
+        for tag in module.tags.import_uv:
```
[the remark this answers is not in the document]
Comment (jvolkman) [reply pycross/private/bzlmod/lock_import.bzl:114]: I think expectations can be set with comments or documentation. Encoding it in the name just means we'll need a naming migration path when the lock file becomes non-experimental.

File: pycross/private/lock_attrs.bzl [hunk 2]
```diff
@@ -92,7 +92,7 @@ RENDER_ATTRS = dict(
     ),
 ) | CREATE_REPOS_ATTRS
 
-PDM_IMPORT_ATTRS = dict(
+_IMPORT_ATTRS = dict(
     lock_file = attr.label(
         doc = "The pdm.lock file.",
```
Comment (jvolkman) [inline pycross/private/lock_attrs.bzl:97]: This comment isn't accurate when used with uv. Options are to extract some common attributes and append PDM- and uv-specific `lock_file` entries, or to just change the doc to not specify a filename.

File: pycross/private/tools/uv_translator.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+    return [Requirement(dep) for dep in deps]
+
+
+def get_optional_dependencies(lock: Dict[str, Any]) -> Dict[str, List[Requirement]]:
+    dep_groups = lock.get("project", {}).get("optional-dependencies", {})
+    return {group: [Requirement(dep) for dep in deps] for group, deps in dep_groups.items()}
+
+
+def get_development_dependencies(lock: Dict[str, Any]) -> Dict[str, List[Requirement]]:
+    dep_groups = lock.get("tool", {}).get("pdm", {}).get("dev-dependencies", {})
```
Comment (jvolkman) [inline pycross/private/tools/uv_translator.py:56]: I assume the `"pdm"` here isn't correct for `uv`, right?

File: pycross/private/tools/uv_translator.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+        type=Path,
+        required=True,
+        help="The path to pyproject.toml.",
+    )
+
+    parser.add_argument(
+        "--lock-file",
+        type=Path,
+        required=True,
+        help="The path to pdm.lock.",
```
Comment (jvolkman) [inline pycross/private/tools/uv_translator.py:302]: `uv.lock`

File: pycross/private/tools/uv_translator.py [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+        required=True,
+        help="The path to the output bzl file.",
+    )
+
+    return parser.parse_args()
+
+
+def collect_and_process_packages(packages_list: list[Dict[str, Any]]) -> Dict[PackageKey, Package]:
+    distinct_packages: Dict[PackageKey, Package] = {}
+    # Pull out all Package entries in a pdm-specific model.
```
Comment (jvolkman) [inline pycross/private/tools/uv_translator.py:355]: uv-specific model

**Review by jvolkman — COMMENTED** [review]
Thanks; this looks fine. I've left a couple of minor comments.

**Comment by jvolkman** [comment]
Thanks!

**Pull request merged**